File: packages/ts/core/src/models/response.ts
import { z } from "zod";

/** Why generation stopped. */
export const FinishReasonSchema = z.enum([
  "stop",
  "length",
  "tool_use",
  "content_filter",
]);
export type FinishReason = z.infer<typeof FinishReasonSchema>;

/** Normalized token accounting. `reasoningTokens` is optional. */
export const TokenUsageSchema = z.object({
  promptTokens: z.number().int().nonnegative(),
  completionTokens: z.number().int().nonnegative(),
  reasoningTokens: z.number().int().nonnegative().optional(),
});
export type TokenUsage = z.infer<typeof TokenUsageSchema>;

/** A tool the model asks the host to run. */
export const ToolInvocationRequestSchema = z.object({
  id: z.string(),
  name: z.string(),
  arguments: z.record(z.string(), z.unknown()),
});
export type ToolInvocationRequest = z.infer<typeof ToolInvocationRequestSchema>;

/** What the host hands back for a tool call on the next turn. */
export const ToolResultSchema = z.object({
  id: z.string(),
  name: z.string(),
  content: z.string(),
});
export type ToolResult = z.infer<typeof ToolResultSchema>;

/** The single normalized outbound shape every adapter produces. */
export const UnifiedResponseSchema = z.object({
  text: z.string(),
  usage: TokenUsageSchema,
  toolCalls: z.array(ToolInvocationRequestSchema),
  finishReason: FinishReasonSchema,
});
export type UnifiedResponse = z.infer<typeof UnifiedResponseSchema>;

File: packages/py/llmorch-core/src/llmorch_core/models/response.py
"""Outbound response models — twin of packages/ts/core/src/models/response.ts."""

from typing import Any, Literal

from pydantic import Field

from .base import CamelModel

FinishReason = Literal["stop", "length", "tool_use", "content_filter"]


class TokenUsage(CamelModel):
    prompt_tokens: int = Field(alias="promptTokens", ge=0)
    completion_tokens: int = Field(alias="completionTokens", ge=0)
    reasoning_tokens: int | None = Field(default=None, alias="reasoningTokens", ge=0)


class ToolInvocationRequest(CamelModel):
    id: str
    name: str
    arguments: dict[str, Any]


class ToolResult(CamelModel):
    id: str
    name: str
    content: str


class UnifiedResponse(CamelModel):
    text: str
    usage: TokenUsage
    tool_calls: list[ToolInvocationRequest] = Field(alias="toolCalls")
    finish_reason: FinishReason = Field(alias="finishReason")
